### PR TITLE
03020 direct token calls

### PR DESCRIPTION
**Pull request by nikolovyanko**
Description:
This PR introduces an alternative way for calling token accounts directly without the need of spinning up the EVM.

-  A new [DirectCallsTxProcessor](https://github.com/hashgraph/hedera-services/blob/bc67246ffbc369a875ffa753daa495fc7dbc5eca/hedera-node/src/main/java/com/hedera/services/contracts/execution/DirectCallsTxProcessor.java) class is introduced, which mimics the taxation of the account and does the gas calculation without calling the EVM.

- A new [TxProcessorUtil](https://github.com/hashgraph/hedera-services/blob/bc67246ffbc369a875ffa753daa495fc7dbc5eca/hedera-node/src/main/java/com/hedera/services/utils/TxProcessorUtil.java) class that wraps the common logic between [DirectCallsTxProcessor](https://github.com/hashgraph/hedera-services/blob/bc67246ffbc369a875ffa753daa495fc7dbc5eca/hedera-node/src/main/java/com/hedera/services/contracts/execution/DirectCallsTxProcessor.java) and [EvmTxProcessor](https://github.com/hashgraph/hedera-services/blob/bc67246ffbc369a875ffa753daa495fc7dbc5eca/hedera-node/src/main/java/com/hedera/services/contracts/execution/EvmTxProcessor.java).

-  Adds a [PrecompileMessage](https://github.com/hashgraph/hedera-services/blob/78532b413228383a3d6db8aceb0d736a0b764a7d/hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/PrecompileMessage.java) class. Since some of the fields from MessageFrame class are needed but not all, the [PrecompileMessage](https://github.com/hashgraph/hedera-services/blob/78532b413228383a3d6db8aceb0d736a0b764a7d/hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/PrecompileMessage.java) replicates them.
- Adds a callHtsDirectly method in [HTSPrecompiledContract](https://github.com/hashgraph/hedera-services/blob/e5905cbcd8ce0df6572ab74708cae02d0ea59a6e/hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java) that is alternative to the compute method used by the EVM approach. 

- Adds a [PrecompileInfoProvider](https://github.com/hashgraph/hedera-services/blob/1d7cbdc558d1e22712961e75c21133327e19bb00/hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/PrecompileInfoProvider.java) interface used as a triggering argument for the different implementations of the [Precompile](https://github.com/hashgraph/hedera-services/blob/78532b413228383a3d6db8aceb0d736a0b764a7d/hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/Precompile.java) interface. With 2 implementations:
    - [EVMPrecompileInfoProvider](https://github.com/hashgraph/hedera-services/blob/1d7cbdc558d1e22712961e75c21133327e19bb00/hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/EVMPrecompileInfoProvider.java) class wrapping a MessageFreame object, used for executing the old logic for token calls with EVM spin up. 
    - [DirectCallsPrecompileInfoProvider](https://github.com/hashgraph/hedera-services/blob/1d7cbdc558d1e22712961e75c21133327e19bb00/hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/DirectCallsPrecompileInfoProvider.java) a class wrapping the newly created [PrecompileMessage](https://github.com/hashgraph/hedera-services/blob/78532b413228383a3d6db8aceb0d736a0b764a7d/hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/PrecompileMessage.java), called for direct token call cases where the EVM does not populate a MessageFrame object.

-  Adds new contracts.enableDirectTokenCalls feature flag. The flag is currently used only in [ContractCallTransitionLogic](https://github.com/hashgraph/hedera-services/blob/78532b413228383a3d6db8aceb0d736a0b764a7d/hedera-node/src/main/java/com/hedera/services/txns/contract/ContractCallTransitionLogic.java) when choosing whether to use the new direct token call logic. 

**Related issue(s):** [3020](https://github.com/hashgraph/hedera-services/issues/3020)

**Checklist**

- [x] Implementation
- [x] Unit tests 
- [x] E2E tests

[Comparison between the direct token call and the previous approach](https://www.notion.so/limechain/Performance-comparison-for-token-call-operations-03020-3686bd2d83eb489da43625011e9c1ef9)

### CODE REVIEW

File: hedera-node/src/main/java/com/hedera/services/context/properties/GlobalDynamicProperties.java [hunk 1]
```diff
@@ -521,6 +523,8 @@ public boolean isHTSPrecompileCreateEnabled() {
 		return enableHTSPrecompileCreate;
 	}
 
+	public boolean isDirectHTSTokenCall(){return enableDirectHTSTokenCalls;}
```
Comment (stoqnkpL) [inline hedera-node/src/main/java/com/hedera/services/context/properties/GlobalDynamicProperties.java:526]: This will be more readable if the method is also called `enableDirectHTSTokenCalls()`

**Review by stoqnkpL — CHANGES_REQUESTED** [review]

File: hedera-node/src/main/java/com/hedera/services/txns/contract/ContractCallTransitionLogic.java [hunk 2]
```diff
@@ -164,6 +181,52 @@ public void doStateTransitionOperation(
...
+		if (properties.shouldEnableTraceability()) {
+			stateChanges = updater.getFinalStateChanges();
+		} else {
+			stateChanges = Map.of();
+		}
+		if (gasAvailable < requiredGas) {
+			//A custom exception or ransactionProcessingResult.failed?
+			isSuccessful = false;
+		}
+		final var gasUsedByTransaction = evmTxProcessor.calculateGasUsedByTokenCallTX(op.getGas(), gasAvailable);
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/txns/contract/ContractCallTransitionLogic.java:205]: We should add logic for refunding accounts for both regular calls and eth calls, as it's done [here](https://github.com/hashgraph/hedera-services/blob/8f12fff2943bbbde02233155bc74156723a59e14/hedera-node/src/main/java/com/hedera/services/contracts/execution/EvmTxProcessor.java#L292)

**Review by IvanKavaldzhiev — CHANGES_REQUESTED** [review]
It's a great start but there is missing logic and needs polishing.

File: hedera-node/src/main/java/com/hedera/services/contracts/execution/EvmTxProcessor.java [hunk 3]
```diff
@@ -360,19 +349,40 @@ private long calculateGasUsedByTX(final long txGasLimit, final MessageFrame init
 		return gasUsedByTransaction;
 	}
 
-	protected long gasPriceTinyBarsGiven(final Instant consensusTime, boolean isEthTxn) {
+	public long calculateGasUsedByTokenCallTX(final long txGasLimit, final long remainingGas) {
+		long gasUsedByTransaction = txGasLimit - remainingGas;
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/contracts/execution/EvmTxProcessor.java:353]: We can simulate the remaining gas calculation by the following logic:

The `remainingGas` value inside the EVM is assigned with the `initialGas` value, which we can calculate. Since we don't spin up the EVM it doesn't consume any gas, so we can just use the `initialGas` as the `remainingGas`.

The `initialGas` is:

`final var gasAvailable = gasLimit - intrinsicGas;`

where:

`final long intrinsicGas = gasCalculator.transactionIntrinsicGasCost(Bytes.EMPTY, contractCreation);`

and gasLimit is the `amount` from contract_call.proto in the case of contract call.

File: hedera-node/src/main/java/com/hedera/services/txns/contract/ContractCallTransitionLogic.java [hunk 4]
```diff
@@ -114,11 +123,11 @@ public void doStateTransitionOperation(
 		var op = contractCallTxn.getContractCall();
 		final var target = targetOf(op);
 		final var targetId = target.toId();
-
 		// --- Load the model objects ---
 		final var sender = accountStore.loadAccount(senderId);
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/txns/contract/ContractCallTransitionLogic.java:127]: As in `EvmTxProcessor`, we can calculate `upfrontGasCost` and check if the sender has enough balance and fail the txn if they don't have enough balance.

**Comment by IvanKavaldzhiev** [comment]
Since this direct token call logic would be complex and replicate a lot of the logic inside `EvmTxProcessor`, probably we can define a separate class (e.g. `EvmTxProcessorSimulator`) and extract a lot of the logic there and clear the `ContractCallTransitionLogic` class. Based on having or not an eth call there might be two different methods using overloading by having different parameters.

File: hedera-node/src/main/java/com/hedera/services/contracts/execution/EvmTxProcessorSimulator.java [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+		}
+
+	}
+
+	private Bytes constructRedirectBytes(Bytes input, long tokenId) {
+		final String TOKEN_CALL_REDIRECT_HEX = "0x618dc65e0000000000000000000000000000000000000";
+		var redirectBytes = Bytes.fromHexString(
+				TOKEN_CALL_REDIRECT_HEX
+						.concat(Long.toHexString(tokenId))
+						.concat(input.toHexString()
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/contracts/execution/EvmTxProcessorSimulator.java:216]: You can use `toUnprefixedHexString()`

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java [hunk 6]
```diff
@@ -607,22 +638,24 @@ protected Bytes computeInternal(final MessageFrame frame) {
...
 			}
 		} catch (final Exception e) {
 			log.warn("Internal precompile failure", e);
 			childRecord = creator.createUnsuccessfulSyntheticRecord(FAIL_INVALID);
 			result = precompile.getFailureResultFor(FAIL_INVALID);
-			addContractCallResultToRecord(childRecord, result, Optional.of(FAIL_INVALID), frame);
+			addContractCallResultToRecord(childRecord, result, Optional.of(FAIL_INVALID), infoProvider);
 		}
+		if (provider.isDirectTokenCall())
+			return splitter(result, childRecord);
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java:658]: You can put { } for better readability.

File: hedera-node/src/main/java/com/hedera/services/txns/contract/ContractCallTransitionLogic.java [hunk 7]
```diff
@@ -128,30 +138,60 @@ public void doStateTransitionOperation(
...
+						op.getGas(),
+						op.getAmount(),
+						callData,
+						txnCtx.consensusTime(),
+						aliasManager.resolveForEvm(receiver.canonicalAddress()),
+						offeredGasPrice,
+						maxGasAllowanceInTinybars,
+						null,
+						targetId.num(),
+						entityAccess.worldLedgers());
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/txns/contract/ContractCallTransitionLogic.java:156]: I think we can add a specific `executeEth()` method specifically for eth txns, as we have separate methods in `evmTxProcessor`. In this way we won't pass `relayer`, `offeredGasPrice` and `maxGasAllowance`. Since we pass here the `offeredGasPrice` and `maxGasAllowance` we are prone to have the side-effect of misleading the `EvmTxnProcessorSimulator` to make gas calculations based on ethereum logic while we have a regular hedera call.

File: hedera-node/src/main/java/com/hedera/services/contracts/execution/EvmTxProcessorSimulator.java [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+					exceptionalHaltReason,
+					stateChanges);
+		}
+
+	}
+
+	private PrecompileMessage constructMessageAndCallPrecompileContract(Account sender, Instant consensusTime, WorldLedgers ledgers,
+																		Bytes payload, long gasAvailable, long value,
+																		long tokenId) {
+		final var redirectBytes = constructRedirectBytes(payload, tokenId);
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/contracts/execution/EvmTxProcessorSimulator.java:280]: Since we implement direct token calls, probably we can think of eliminating the use of redirect bytecode that we artificially build. If it's easily achievable we can try to directly invoke the correct ERCPrecompile (e.g. `balanceOf`, `totalSupply`, `transferFrom`). This would make even a better "shortcut" to execute the ERC precompile logic and reduce the complexity.

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/DirectCallsInfoProvider.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+public record DirectCallsInfoProvider(PrecompileMessage precompileMessage) implements InfoProvider {
+	@Override
+	public Wei getValue() {
+		return precompileMessage.getValue();
+	}
+
+	@Override
+	public MessageFrame messageFrame() {
+		return null;
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/DirectCallsInfoProvider.java:39]: To avoid returning null directly, wouldn't it be better if we have `Optional<MessageFrame>`?

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/utils/KeyActivationUtils.java [hunk 10]
```diff
@@ -69,14 +72,20 @@ private KeyActivationUtils() {
...
-		final var recipient = aliases.resolveForEvm(frame.getRecipientAddress());
-		final var sender = aliases.resolveForEvm(frame.getSenderAddress());
+		//logic for direct token account calls
+		if(provider instanceof DirectCallsInfoProvider){
+			final var message =((DirectCallsInfoProvider) provider).precompileMessage();
+			return activationTest.apply(false, target, message.getSenderAddress(), ledgers);
+		}
+		final var frame = ((EVMInfoProvider)provider).messageFrame();
+		final var recipient = aliases.get().resolveForEvm(frame.getRecipientAddress());
+		final var sender = aliases.get().resolveForEvm(frame.getSenderAddress());
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/utils/KeyActivationUtils.java:89]: There might be a case (even if not present now) where we can have an `EVMInfoProvider` but have not set `ContractAliases` optional value. In this case we would have `NullPointerException`. We can think of a graceful way to handle this case.

File: hedera-node/src/main/java/com/hedera/services/contracts/execution/EvmTxProcessor.java [hunk 11]
```diff
@@ -323,41 +278,13 @@ protected TransactionProcessingResult execute(
...
-		gasUsedByTransaction = Math.max(gasUsedByTransaction, txGasLimit - txGasLimit * maxRefundPercent / 100);
-
-		return gasUsedByTransaction;
+		final var isCompletedSuccess = initialFrame.getState() == MessageFrame.State.COMPLETED_SUCCESS;
+		return TxProcessorUtil.buildTransactionResult(
+				isCompletedSuccess, initialFrame.getLogs(),
+				updater.getSbhRefund(), gasUsedByTransaction,
+				gasPrice, mirrorReceiver, initialFrame.getOutputData(),
+				initialFrame.getRevertReason(), initialFrame.getExceptionalHaltReason(),
+				stateChanges);
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/contracts/execution/EvmTxProcessor.java:287]: You can minimise the number of the arguments a lot if you do the following:

- Pass the initialFrame/messageFrame itself and then internally get the needed fields
- Instead of passing `isCompletedSuccess` you can also determine its value internally inside `buildTransactionResult`

Comment (nikolovyanko) [reply hedera-node/src/main/java/com/hedera/services/contracts/execution/EvmTxProcessor.java:287]: Unfortunately, MessageFrame and PrecompileMessage classes don't share a common parent, and can't be passed to the same method as arguments.

File: hedera-node/src/main/java/com/hedera/services/utils/TxProcessorUtil.java [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+					gasUsedByTransaction,
+					sbhRefund,
+					gasPrice,
+					revertReason,
+					exceptionalHaltReason,
+					stateChanges);
+		}
+	}
+
+
```
Comment (stoqnkpL) [inline hedera-node/src/main/java/com/hedera/services/utils/TxProcessorUtil.java:221]: nit: empty line

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/InfoProvider.java [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+ * ‍
+ */
+
+import org.apache.tuweni.bytes.Bytes;
+import org.hyperledger.besu.datatypes.Address;
+import org.hyperledger.besu.datatypes.Wei;
+import org.hyperledger.besu.evm.frame.MessageFrame;
+import org.hyperledger.besu.evm.log.Log;
+
+public interface InfoProvider {
```
Comment (stoqnkpL) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/InfoProvider.java:29]: `InfoProvider` sounds very general, it could be anything, can we give a more specific name to this class like `PrecompileContext` or `PrecompileInfoProvider`

File: hedera-node/src/main/java/com/hedera/services/contracts/execution/DirectCallsTxProcessor.java [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+			TxProcessorUtil.chargeGas(gasCost, upfrontCost, mutableSender);
+		} else {
+			allowanceCharged = TxProcessorUtil.chargeForEth(
+					userOfferedGasPrice, gasCost, maxGasAllowanceInTinybars, mutableSender,
+					mutableRelayer, Wei.ZERO, gasPrice, gasLimit, value);
+		}
+
+		PrecompileMessage message = PrecompileMessage.builder()
+				.setLedgers(ledgers)
+				.setSenderAddress(sender.canonicalAddress())
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/contracts/execution/DirectCallsTxProcessor.java:112]: ⚠️ If the sender is an EOA (i.e., an auto-created account with an ECDSA key) then this will blow up any precompile that uses `senderAddress`. 

For example, I've pushed a branch with EET [here](https://github.com/hashgraph/hedera-services/blob/add-failing-eet/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/hapi/ContractCallLocalSuite.java#L227) that pays for an ERC-20 `approve()` with an EOA; not only does the precompile incorrectly get `INVALID_ACCOUNT_ID`, the record includes,
```
sender_id {
  shardNum: -70672035
  realmNum: 8062295507940796406
  accountNum: 4428982008424832196
}
```

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/utils/KeyActivationUtils.java [hunk 15]
```diff
@@ -58,25 +61,31 @@ private KeyActivationUtils() {
...
 	 * @return whether the implied key is active
 	 */
 	public static boolean validateKey(
-			final MessageFrame frame,
+			final PrecompileInfoProvider provider,
 			final Address target,
 			final KeyActivationTest activationTest,
 			final WorldLedgers ledgers,
-			final ContractAliases aliases
+			final Optional<ContractAliases> aliases
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/utils/KeyActivationUtils.java:76]: This seems workable, but I would greatly prefer to revert the changes to this method, and create a new method especially for direct calls,

```
public static boolean validateKeyForDirectCall(
    final DirectCallsPrecompileInfoProvider directProvider,
    final Address target.
    final KeyActivationTest activationTest,
    final WorldLedgers ledger
) {
```

**Review by tinker-michaelj — CHANGES_REQUESTED** [review]
A lot of very nice work here! 🥂 Appreciate the effort in avoiding duplication with `EvmTxProcessor`. A few comments:

 - I think we should avoid making the semantics of `KeyActivationUtils.validateKey()` any more complicated, so had a comment there.
 - There's a blocking issue with use of canonical `senderAddress` for a direct call (I pushed a branch with a failing EET [here](https://github.com/hashgraph/hedera-services/blob/add-failing-eet/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/hapi/ContractCallLocalSuite.java#L230)).
 - It will be a bit tiresome, but imo we should write at least basic EET's of that type that do direct calls for all supported ERC-20 and ERC-721 methods. 😓

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/utils/KeyActivationUtils.java (hunk 15 above)
Comment (nikolovyanko) [reply hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/utils/KeyActivationUtils.java]: Thanks for the detailed review and comments. However, I'm not sure that I understood this one correctly. Does the last commit simplify the logic for validateKey enough, or the aim here is to explicitly check every precompile impl. for directToken logic and call either validateKeyForDirectCall or validateKey(MessageFrame...)?

Comment (tinker-michaelj) [reply hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/utils/KeyActivationUtils.java]: Np! I like the step you took...could we go further by adding a method to `PrecompileInfoProvider`,
```
   ...
   @Nullable
   ContractAliases aliases();
```
...and then have,
```
	public static boolean validateKey(
			final PrecompileInfoProvider precompileInfoProvider,
			final Address target,
			final KeyActivationTest activationTest,
			final WorldLedgers ledgers,
			@Nullable final ContractAliases contractAliases
	) {
		if (precompileInfoProvider.isDirectTokenCall()) {
			return validateKeyForDirectCall((DirectCallsPrecompileInfoProvider) precompileInfoProvider,
					target, activationTest, ledgers);
		}
                validateTrue(contractAliases != null, FAIL_INVALID);
                return validateKeyForCallWithMessageFrame(frame, target, activationTest, ledgers, aliases);
	}
```
Imo this would make both this method _and_ the client code that calls it cleaner.

File: hedera-node/src/main/java/com/hedera/services/utils/TxProcessorUtil.java [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+					sbhRefund,
+					gasPrice,
+					revertReason,
+					exceptionalHaltReason,
+					stateChanges);
+		}
+	}
+
+	private TxProcessorUtil() {
+		throw new IllegalStateException("Utility class");
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/utils/TxProcessorUtil.java:206]: If you change this to throw an `UnsupportedOperationException` (e.g. [here](https://github.com/hashgraph/hedera-services/blob/master/hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/AbiConstants.java#L29)), and then add `TxProcessorUtil.class` to [this list](https://github.com/hashgraph/hedera-services/blob/master/hedera-node/src/test/java/com/hedera/services/utils/UtilsConstructorTest.java#L100), you get automatic coverage of such constructors.

File: hedera-node/src/main/java/com/hedera/services/contracts/execution/DirectCallsTxProcessor.java [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+				gasPrice, mirrorReceiver, message.getHtsOutputResult(),
+				message.getRevertReason(), message.getExceptionalHaltReason(),
+				Map.of());
+	}
+
+	private void callHtsPrecompile(PrecompileMessage message) {
+		htsPrecompiledContract.callHtsPrecompileDirectly(message);
+		final var gasRequirement = message.getGasRequired();
+
+		if (message.getGasRemaining() < gasRequirement) {
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/contracts/execution/DirectCallsTxProcessor.java:147]: Afaict this block is an exact duplicate of [here](https://github.com/hashgraph/hedera-services/blob/master/hedera-node/src/main/java/com/hedera/services/contracts/execution/HederaMessageCallProcessor.java#L91); if we extract that to a static method and call here, we can fix missing code coverage on L#156.

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/PrecompileMessage.java [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+		}
+
+
+		public PrecompileMessage build() {
+			return new PrecompileMessage(this.ledgers, this.senderAddress, this.value, this.consensusTime,
+					this.gasRemaining, this.inputData, this.tokenID);
+		}
+	}
+
+	public enum State {
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/PrecompileMessage.java:216]: Imo we don't gain anything by redefining the `MessageFrame.State` enumeration here; would prefer to just re-use that type.

**Review by tinker-michaelj — COMMENTED** [review]
Appreciate the `KeyActivationUtils` clean up! 🙏 Had a few other minor comments.

I think we need more end-to-end tests, especially negative cases, to be fully confident. Are the EET tests planned in a separate PR?

Also, I was thinking and had a question: Is our goal to make the `ContractCallResult` from a direct token call look exactly like the `ContractCallResult` of a corresponding ERC-20/721 contract?

Right now, these two results will be _quite_ different, because a direct call's `ContractCallResult` is in a child record.

**Comment by shemnon** [comment]
Meta-review.  We are going from something that works, and is simple, to something else that works, but is not as simple. The  justification for that is performance.

Do we have any before and after performance numbers?  2.8K line delta for a presumed efficiency gain feels too risky.  I would need to see what the real gains are, so we can see if the gains are worth this much code impact.

**Comment by stoqnkpL** [comment]
> Meta-review. We are going from something that works, and is simple, to something else that works, but is not as simple. The justification for that is performance.
> 
> Do we have any before and after performance numbers? 2.8K line delta for a presumed efficiency gain feels too risky. I would need to see what the real gains are, so we can see if the gains are worth this much code impact.

We summarized the performance analysis results on this [notion page](https://limechain.notion.site/Performance-comparison-for-token-call-operations-03020-06c4e344b7cf44beb5363dfe12b4ab4a). It was done on one of our machines with a locally ran node.

File: hedera-node/src/main/java/com/hedera/services/contracts/execution/DirectCallsTxProcessor.java (hunk 17 above)
Comment (nikolovyanko) [reply hedera-node/src/main/java/com/hedera/services/contracts/execution/DirectCallsTxProcessor.java:147]: It would be great to remove the duplicated code here, unfortunately, MessageFrame and PrecomileMessage are not related by a common parent. Hence, such a method signature can't be applied. What comes to my mind as a possible solution is to introduce an additional model class to be passed as a first argument, instead of MessageFrame or an additional method with PrecompileMessage argument and a private method for the common logic shared by both methods. But I'm not sure if these approaches are appropriate here? I'm open to suggestions! 
:raised_hands:

**Comment by shemnon** [comment]
I don't find those numbers persuasive. The performance gains may be there, but this is not the report that justifies it.

The two numbers pulled out (Average Memory used and Total Number of Classes) are at best secondary indicators.  Because we are running the service on data center nodes this level of delta is not that impactful and where we would feel the pain.  A more relevant number would be the TPS that can be sustained or the total time to execute a fixed number of calls that would take about a minute in the pre-existing case.

These tests also don't consider that the TXes will be intermixed with regular old smart contract calls in addition to HTS calls.

What I would consider more persuasive is a performance test with 1000 HTS calls intermixed with 1000 regular smart contract calls (against an ERC-20 contract), and then gauge the total time to execute.  Perhaps both performing transfers on a fixed schedule, like an air drop.

**Comment by stoqnkpL** [comment]
I ran [this performance EET](https://github.com/hashgraph/hedera-services/blob/9dd0e971af9ae79dce2e84bebcc0dd56085273c3/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/ERCDirectTokenCallSuite.java#L1006) on my local machine. Here's what the test does:
- increases the throttle limits for ContractCalls to 5k
- disables throttling by gas
- creates an HTS token
- executes 3000 ContractCall operations to the token address with the ERC20ABI for transfer, each call transferring 1 token unit
- extracted execution times from MainNetStats0.csv - ContractCall Operations executed column.

I ran the test 3 times on a continuously running environment. I assume each row from the table is approximately 1s of execution time.

Results from the master branch:
1st run: 6s.
2nd run: 5s.
3rd run: 4s.

Results from the 03020_direct_token_calls branch:
1st run: 4s.
2nd run: 4s.
3rd run: 4s.

It looks like there is a small performance improvement with the direct calls approach. I'm not sure if there's any way to get more detailed results.
Attaching both csv files for reference; 
[MainNetStats0-direct-calls-branch.csv](https://github.com/hashgraph/hedera-services/files/9179475/MainNetStats0-direct-calls-branch.csv)
[MainNetStats0-master.csv](https://github.com/hashgraph/hedera-services/files/9179476/MainNetStats0-master.csv)

Based on these findings and further discussions with the team, we decided that the performance gains were not big enough to justify such a big change.
